### ISSUE TITLE
About the acc

### ISSUE THREAD

**Issue by Jiacheng8** [LINs-lab/RDED#4]
Hi, first of all, thanks for the great work you have done!

I have been working with the CIFAR-100 dataset using your script with the following command:
```
python ./main.py \
--subset "cifar100" \
--arch-name "resnet18_modified" \
--factor 1 \
--num-crop 5 \
--mipc 300 \
--ipc 50 \
--stud-name "resnet18_modified" \
--re-epochs 300

```
However, I noticed that the accuracy I obtained on the test set is 54.25%, which is significantly lower than the 62.6 ± 0.1% reported in your paper. I would greatly appreciate it if you could help clarify any potential reasons for this discrepancy.

Additionally, I wanted to ask whether the labels for the CIFAR-100 dataset in your experiment were the same as the original dataset. Also, could you confirm if the teacher model's performance on the training set and test set was similar to the following:

Train set: loss = 0.603734, Top-1 error = 17.14%, Top-5 error = 3.16%
Test set: loss = 1.890325, Top-1 error = 41.41%, Top-5 error = 16.46%
Thank you very much for your assistance, and I look forward to your response!

**Comment by Jiacheng8** [comment]
It is the label issue, the writer's result is fully correct

**Comment by ai23resch04001** [comment]
Hi @Jiacheng8 , could you please elaborate what did you  mean by the "label issue". I am facing a similar accuracy discrepancy for IPC=50 and would like to understand what I am doing wrong.

**Comment by Jiacheng8** [comment]
Hi, the label issue refers to the sequence of the labels; for example, if you directly use the cifar100 dataset downloaded from torchvision, its label sequence is different from the label sequence used in the pretrained model, and thus result in performance degradation, I have a code which can transfer the original downloaded torchvision dataset into aligned label sequence used in the pretrained model for CIFAR-100 dataset.

```python
import torchvision
import os
from PIL import Image
import numpy as np
import pickle
import os
from tqdm import trange
from os.path import join
import torchvision
import argparse

def format_int_to_str(number):
    return "{:05}".format(number)

def my_mkdirs(path):
    if not os.path.exists(path):
        os.makedirs(path)


def unpickle(file):
    with open(file, 'rb') as fo:
        dict = pickle.load(fo, encoding='latin1')
    return dict

parser = argparse.ArgumentParser("generate_cifa_dataset")
parser.add_argument('--save-dir', type=str, default='/l/users/jiacheng.cui/Nas_DD_data/dataset',
                    help='The directory to save the CIFAR-100 dataset')
parser.add_argument('--dataset_save_dir', type=str, default='/l/users/jiacheng.cui/Nas_DD_data/dataset/cifar100',
                    help='The directory to save the image dataset')
args = parser.parse_args()

if __name__ == '__main__':
    save_dir = args.save_dir
    src_dir = join(save_dir, 'cifar-100-python')
    dst_dir = args.dataset_save_dir
    os.makedirs(args.save_dir, exist_ok=True)
    trainset = torchvision.datasets.CIFAR100(
        root=save_dir,
        train=True,
        download=True
    )

    testset = torchvision.datasets.CIFAR100(
        root=save_dir,
        train=False,
        download=True
    )

    print(f"CIFAR-100 is downloaded to: {save_dir}")
    
    meta = unpickle(join(src_dir, 'meta')) # KEYS: {'fine_label_names', 'coarse_label_names'}
    my_mkdirs(dst_dir)

    for data_set in ['train', 'test']:
        print('Unpickling {} dataset......'.format(data_set))
        data_dict = unpickle(join(src_dir, data_set)) # KEYS: {'filenames', 'batch_label', 'fine_labels', 'coarse_labels', 'data'}
        my_mkdirs(join(dst_dir, data_set))

        for fine_label_name in meta['fine_label_names']:
            my_mkdirs(join(dst_dir, data_set, fine_label_name))

        for i in trange(data_dict['data'].shape[0]):
            img = np.reshape(data_dict['data'][i], (3, 32, 32))
            i0 = Image.fromarray(img[0])
            i1 = Image.fromarray(img[1])
            i2 = Image.fromarray(img[2])
            img = Image.merge('RGB', (i0, i1, i2))
            img.save(join(dst_dir, data_set, meta['fine_label_names'][data_dict['fine_labels'][i]], data_dict['filenames'][i]))

    print('All done.')
    
import os
import torchvision

# CIFAR-100 标签与类别名映射
trainset = torchvision.datasets.CIFAR100(root=save_dir, train=True, download=True)
label_to_class_mapping = {class_name: i for i, class_name in enumerate(trainset.classes)}

def rename_subfolders(root_dir):
    # 循环遍历根目录下的所有子文件夹
    for subfolder in os.listdir(root_dir):
        subfolder_path = os.path.join(root_dir, subfolder)
        if os.path.isdir(subfolder_path):
            # 查找子文件夹名对应的 CIFAR-100 标签
            if subfolder in label_to_class_mapping:
                label = label_to_class_mapping[subfolder]
                # 构造新的文件夹名称
                new_folder_name = str(format_int_to_str(label))
                new_folder_path = os.path.join(root_dir, new_folder_name)
                # 重命名文件夹
                os.rename(subfolder_path, new_folder_path)
            else:
                print(f"Warning: Subfolder '{subfolder}' not found in CIFAR-100 class mapping")


# 使用该函数遍历并重命名指定路径下的子文件夹中的文件
root_directory = os.path.join(dst_dir,"train")  # 替换为实际的路径
rename_subfolders(root_directory)

root_directory = os.path.join(dst_dir,"test")  # 替换为实际的路径
rename_subfolders(root_directory)
```

This code should be ready to run, but you need to change some of the argument

**Comment by ai23resch04001** [comment]
Hi @Jiacheng8 ,
Thanks a lot for the script. Yes, that seems to work out and the accuracy values are now matching the values mentioned in the paper.
Could you please share a similar script for Tiny ImageNet and ImageNet-1K , if it is available with you?

Thanks again.

**Comment by Wenxin-Rong** [comment]
Hello Jiacheng, thank you very much for sharing the script. 
Is there any further modification needed for TinyImageNet or ImageNet-1K? I have resolved the label issues and evaluated the pretrained model with the modified labels. The model's performance is good, indicating that the pretrained model itself is not problematic. However, the accuracy is still about 10% lower than reported in the paper.
Thank you again.

**Comment by Jiacheng8** [comment]
For your question, I didn't do any further modification for ImageNet-1k and Tiny-ImageNet. 

For Tiny-ImageNet, there should be two label sequences, so if one doesn't work for you, you should try to use the alternative ones. I did reproduce the result, the only un-matching result is Tiny-ImageNet Ipc=50 on ResNet-18 as student model, the acc gap is about 5%, I found this is ok, since the selection of patches for RDED has large randomness, maybe they select a super-good patch. But 10% lower indicates something is abnormal.

**Comment by sp12138** [comment]
@Wenxin-Rong @Jiacheng8 @ai23resch04001:

We sincerely apologize for the reproduction issue you encountered. The problem appears to stem from two critical aspects:

1. Please ensure that you are using our pre-trained models.
2. Verify that your raw dataset storage format strictly adheres to our specifications (refer to: https://github.com/LINs-lab/RDED/tree/main/README.md).

To assist you further, we have provided comprehensive documentation for dataset preparation (available at: https://github.com/LINs-lab/RDED/tree/main/prepare).

Moreover, we have included a dedicated script for executing RDED with Tiny-ImageNet (Ipc=50) using ResNet-18 as the student model (accessible at: https://github.com/LINs-lab/RDED/tree/main/scripts).

We deeply regret the delay in our response and appreciate your patience. Should you have any additional questions, please don't hesitate to reach out.
